packages/postgres-database/src/simcore_postgres_database/utils_tags.py

```diff
@@ -222,44 +227,102 @@ async def delete(
         async with transaction_context(self.engine, connection) as conn:
             deleted = await conn.scalar(stmt_delete)
             if not deleted:
-                msg = f"Could not delete {tag_id=}. Not found or insuficient access."
-                raise TagOperationNotAllowedError(msg)
+                raise TagOperationNotAllowedError(
+                    operation="delete", tag_id=tag_id, user_id=user_id
+                )
 
     #
     # ACCESS RIGHTS
     #
 
-    async def create_access_rights(
+    async def has_access_rights(
         self,
         connection: AsyncConnection | None = None,
         *,
-        user_id: int,
+        caller_id: int,
         tag_id: int,
-        group_id: int,
-        read: bool,
-        write: bool,
-        delete: bool,
-    ):
-        raise NotImplementedError
+        read: bool = False,
+        write: bool = False,
+        delete: bool = False,
+    ) -> bool:
+        async with pass_or_acquire_connection(self.engine, connection) as conn:
+            group_id_or_none = await conn.scalar(
+                has_access_rights_stmt(
+                    tag_id=tag_id,
+                    caller_user_id=caller_id,
+                    read=read,
+                    write=write,
+                    delete=delete,
+                )
+            )
+            return bool(group_id_or_none)
 
-    async def update_access_rights(
+    async def list_access_rights(
+        self,
+        connection: AsyncConnection | None = None,
+        *,
+        tag_id: int,
+    ) -> list[TagAccessRightsDict]:
+        async with pass_or_acquire_connection(self.engine, connection) as conn:
+            result = await conn.execute(list_tag_group_access_stmt(tag_id=tag_id))
+            return [
+                TagAccessRightsDict(
+                    tag_id=row.tag_id,
+                    group_id=row.group_id,
+                    read=row.read,
+                    write=row.write,
+                    delete=row.delete,
+                )
+                for row in result.fetchall()
+            ]
+
+    async def create_or_update_access_rights(
         self,
         connection: AsyncConnection | None = None,
         *,
-        user_id: int,
         tag_id: int,
         group_id: int,
+        # access-rights
         read: bool,
         write: bool,
         delete: bool,
-    ):
-        raise NotImplementedError
+    ) -> TagAccessRightsDict:
+        # NOTE that there is no reference to the caller which implies
+        # that we assume that the rigths to create or update these access_rights
+        # have been checked (via has_access_rights) before calling this function
```

Comment:
Q: does it not make more sense for this function to be decorated with the function to check access rights instead?

Reply:
If what you are referring to is `@permission_required` I would just remove the comment. Does not bring much.

Reply:
The access rights are checked at different layers. role-based access is indeed checked with the decorator you refer but dynamic access rights (i.e. those defined in `tags_access_rights`) are defined at the service layer using `_validate_tag_sharing_permissions` (BTW this should be perhaps done with a decorator or via dependency injection!